test/bench.cpp

```diff
@@ -1,90 +1,392 @@
-  }
-
-  virtual void on_gate( const std::vector<std::string>& inputs, const std::string& output, const std::string& type ) const override
-  {
-    gate_lines.emplace_back( inputs, output, type );
-    ++_stats.number_of_lines;
-  }
-
-  virtual void on_assign( const std::string& input, const std::string& output ) const override
-  {
-    (void)input;
-    (void)output;
-    ++_stats.number_of_lines;
-  }
-
-  bench_statistics& _stats;
-  mutable std::vector<std::tuple<std::vector<std::string>,std::string,std::string>> gate_lines;
+
+    explicit bench_statistics_reader(bench_statistics &stats)
+            : _stats(stats) {
+    }
+
+    virtual void on_input(const std::string &name) const override {
+        (void) name;
+        std::vector<std::string> input;
+        addEdge(_stats, input, "input", name);
+        ++_stats.number_of_inputs;
+        ++_stats.number_of_edges;
+    }
+
+    virtual void on_output(const std::string &name) const override {
+        (void) name;
+        std::vector<std::string> input;
+        input.emplace_back(name);
+        addEdge(_stats, input, "output");
+        ++_stats.number_of_outputs;
+        ++_stats.number_of_edges;
+    }
+
+    virtual void on_dff_input(const std::string &input) const override {
+        (void) input;
+    }
+
+    virtual void on_dff(const std::string &input, const std::string &output) const override {
+        (void) input;
+        (void) output;
+        std::vector<std::string> inp;
+        inp.emplace_back(input);
+        addEdge(_stats, inp, "dff", output);
+        ++_stats.number_of_dffs;
+        ++_stats.number_of_edges;
+    }
+
+    virtual void
+    on_gate(const std::vector<std::string> &inputs, const std::string &output, const std::string &type) const override {
+        gate_lines.emplace_back(inputs, output, type);
+        addEdge(_stats, inputs, type, output);
+        ++_stats.number_of_lines;
+        ++_stats.number_of_edges;
+    }
+
+    virtual void on_assign(const std::string &input, const std::string &output) const override {
+        (void) input;
+        (void) output;
+        std::vector<std::string> inp;
+        inp.emplace_back(input);
+        addEdge(_stats, inp, "dff", output);
+        ++_stats.number_of_lines;
+        ++_stats.number_of_edges;
+    }
+
+    bench_statistics &_stats;
+    mutable std::vector<std::tuple<std::vector<std::string>, std::string, std::string>> gate_lines;
 }; /* bench_statistics_reader */
 
+
 static void
 dump_statistics(FILE *f, const bench_statistics &st) {
-  fprintf(f, "inputs: %u, outputs: %u, num ddfs: %u, num lines: %u\n",
-      st.number_of_inputs,
-      st.number_of_outputs,
-      st.number_of_dffs,
-      st.number_of_lines);
+    fprintf(f, "inputs: %u, outputs: %u, num ddfs: %u, num lines: %u\n",
+            st.number_of_inputs,
+            st.number_of_outputs,
+            st.number_of_dffs,
+            st.number_of_lines);
 }
 
+
 int
 main(int argc, char *argv[]) {
```

Comment:
It is okay for `main` to contain init, event loop and shutdown as is, but anything related to data preparation or application logic must be dedicated procedures. Actually, `main` may look like this:
```
int main(int argc, char **argv) {
    init();
    data = load_data(argc, argv);
    while (!exit) {
        draw_data(data);
        handle_inputs();
    }
    shutdown();
}
```